source/common/http/conn_manager_impl.cc

```diff
@@ -756,37 +756,36 @@ void ConnectionManagerImpl::ActiveStream::resetIdleTimer() {
 
 void ConnectionManagerImpl::ActiveStream::onIdleTimeout() {
   connection_manager_.stats_.named_.downstream_rq_idle_timeout_.inc();
-  // If headers have not been sent to the user, send a 408.
-  if (responseHeaders().has_value()) {
-    // TODO(htuch): We could send trailers here with an x-envoy timeout header
-    // or gRPC status code, and/or set H2 RST_STREAM error.
-    filter_manager_.streamInfo().setResponseCodeDetails(
-        StreamInfo::ResponseCodeDetails::get().StreamIdleTimeout);
-    connection_manager_.doEndStream(*this);
-  } else {
-    // TODO(mattklein) this may result in multiple flags. This Ok?
-    filter_manager_.streamInfo().setResponseFlag(StreamInfo::ResponseFlag::StreamIdleTimeout);
-    sendLocalReply(Http::Code::RequestTimeout, "stream timeout", nullptr, absl::nullopt,
-                   StreamInfo::ResponseCodeDetails::get().StreamIdleTimeout);
-  }
+  // TODO(mattklein) this may result in multiple flags. This Ok?
+  filter_manager_.streamInfo().setResponseFlag(StreamInfo::ResponseFlag::StreamIdleTimeout);
+  sendLocalReply(filter_manager_.remoteDecodeComplete() ? Http::Code::GatewayTimeout
```

Comment:
might be worth a temporary helper function especially with  runtime guard and use below

Reply:
I am adding a helper function visible by ActiveStream 
That means the router filter does not use this helper function because router cannot access `FilterManager::remoteDecodeComplete()` anyway
Let me know if that work for you

Reply:
Alternatively, promote maybeRequestTimeoutCode(filter_manager_)  to member function FilterManager::maybeRequestTimeoutCode()

Reply:
choose maybeRequestTimeoutCode(bool) as your comment https://github.com/envoyproxy/envoy/pull/20026#discussion_r824105159